src/Microsoft.AspNet.Routing/RouteCollection.cs

```diff
@@ -197,7 +198,35 @@ public virtual string GetVirtualPath(VirtualPathContext context)
             }
         }
 
-        private void EnsureLogger(HttpContext context)
+        private string NormalizeVirtualPath(String url)
+        {
+            if (string.IsNullOrEmpty(url)) return url;
+
+            if (_options.LowercaseUrls)
+            {
+                var indexOfSeparator = url.IndexOfAny(new char[] { '?', '#' });
+
+                // No query string, lowercase the url
+                if (indexOfSeparator == -1)
+                {
+                    return url.ToLowerInvariant();
+                }
+                else
+                {
+                    var lowercaseUrl = url.Substring(0, indexOfSeparator);
+                    var queryString = url.Substring(indexOfSeparator);
+
+                    lowercaseUrl = lowercaseUrl.ToLowerInvariant();
+
+                    // queryString will contain the delimiter ? or # as the first character, so it's safe to append.
+                    return lowercaseUrl + queryString;
+                }
+            }
+
+            return url;
+    }
```

Comment:
(I don't see the formatting fix in the PR, so I assume it's in your local branch.)